source/Irrlicht/COpenGLDriver.cpp

```diff
@@ -2151,73 +2150,9 @@ void COpenGLDriver::SAuxContext::flushState(GL_STATE_BITS stateBits)
-        GLsizeiptr* const ssbosizes = ASSIGN_MEM(ssbosizes);
-        UPDATE_OFFSET(ssbosizes);
-
-        GLuint* const texnames = ASSIGN_MEM(texnames);
-        UPDATE_OFFSET(texnames);
-        GLuint* const smplrnames = ASSIGN_MEM(smplrnames);
-        UPDATE_OFFSET(smplrnames);
-
-        GLuint* const imgnames = ASSIGN_MEM(imgnames);
-#undef UPDATE_OFFSET
-#undef ASSIGN_MEM
-
-        //unbind previous descriptor sets
-        bool needRebind = prevPipeline != nextState.pipeline;//TODO could optimize to checking pipeline layouts or even pipeline layouts compatibility (that would probably result in needResult being true less often)
-        {
-            constexpr GLuint invalid = ~static_cast<GLuint>(0u);
-            std::pair<GLuint, GLsizei> ubos{invalid,0}, ssbos{invalid,0}, textures{invalid,0}, textureImages{invalid,0};
-            for (uint32_t i = 0u; i < video::IGPUPipelineLayout::DESCRIPTOR_SET_COUNT; ++i)
-            {
-                if (!prevPipeline)
-                    break;
-
-                const auto& first_count = static_cast<const COpenGLPipelineLayout*>(prevPipeline->getLayout())->getMultibindParamsForDescSet(i);
-                const auto& multibind_params = currentState.descriptorsParams.descSets[i]->getMultibindParams();
-                if (needRebind || (nextState.descriptorsParams.descSets[i] != currentState.descriptorsParams.descSets[i]))
-                {
-                    needRebind = true;
-#define OVERRIDE_FIRST(resname) if (resname.first == invalid) { resname.first = first_count.resname.first; }
-                    OVERRIDE_FIRST(ubos);
-                    ubos.second += first_count.ubos.count;
-                    OVERRIDE_FIRST(ssbos);
-                    ssbos.second += first_count.ssbos.count;
-                    OVERRIDE_FIRST(textures);
-                    textures.second += first_count.textures.count;
-                    OVERRIDE_FIRST(textureImages);
-                    textureImages.second += first_count.textureImages.count;
-#undef OVERRIDE_FIRST
-                }
-            }
-            extGlBindBuffersRange(GL_UNIFORM_BUFFER, ubos.first, ubos.second, ubonames, ubooffsets, ubosizes);
-            extGlBindBuffersRange(GL_SHADER_STORAGE_BUFFER, ssbos.first, ssbos.second, ssbonames, ssbooffsets, ssbosizes);
-            extGlBindTextures(textures.first, textures.second, texnames, nullptr); //targets=nullptr: assuming ARB_multi_bind (or GL>4.4) is always available
-            extGlBindSamplers(textures.first, textures.second, smplrnames);
-            extGlBindImageTextures(textureImages.first, textureImages.second, imgnames, nullptr); //formats=nullptr: assuming ARB_multi_bind (or GL>4.4) is always available
-        }
-
         //bind new descriptor sets
-        needRebind = prevPipeline != nextState.pipeline;
+        int64_t newUboCount = 0u, newSsboCount = 0u, newTexCount = 0u, newImgCount = 0u;
+        bool needRebind = prevPipeline != nextState.pipeline;
```

Comment:
wait, what why?

Reply:
its wrong... again

Reply:
see #376

Reply:
idk what i meant by this really. Now i simplified it to what you wrote here https://github.com/buildaworldnet/IrrlichtBAW/pull/358#discussion_r333204559 but optimized a little (if prev and curr pipeline layouts are compatible for set N and `currState.set[N]==nextState.set[N]` then binding set N would be redundant)

Reply:
We need to talk this over in person (Call) and come to some sort of a conclusion in #376

Reply:
resolved